File: solutions/csharp/house/2/House.cs
using System.Text;

public static class House
{
    public static string Recite(int verseNumber)
    {
        var verses = new List<string>
        {
            "the house",
            "malt that lay in ",
            "rat that ate ",
            "cat that killed ",
            "dog that worried ",
            "cow with the crumpled horn that tossed ",
            "maiden all forlorn that milked ",
            "man all tattered and torn that kissed ",
            "priest all shaven and shorn that married ",
            "rooster that crowed in the morn that woke ",
            "farmer sowing his corn that kept ",
            "horse and the hound and the horn that belonged to ",
        };

        // if (verseNumber > 1)
        // {
            var allVerses = new StringBuilder();
            for (var i = verseNumber - 1; i >= 1; i--)
            {
                allVerses.Append(verses[i]);
                allVerses.Append("the ");
            }
            return "This is the " + allVerses.ToString() + "house that Jack built.";
        // }
        // else
        // {
        //     return "This is " + verses[verseNumber - 1] + " that Jack built.";
        // }
    }

    public static string Recite(int startVerse, int endVerse)
    {
        var rhyme = new StringBuilder();
        for (var i = startVerse; i <= endVerse; i++)
        {
            rhyme.Append(Recite(i));
            rhyme.Append('\n');
        }
        return rhyme.ToString().Trim();
    }
}

File: solutions/elixir/house/2/lib/house.ex
defmodule House do

  @actors [
    ["the malt", "lay"],
    ["the rat", "ate"],
    ["the cat", "killed"],
    ["the dog", "worried"],
    ["the cow with the crumpled horn", "tossed"],
    ["the maiden all forlorn", "milked"],
    ["the man all tattered and torn", "kissed"],
    ["the priest all shaven and shorn", "married"],
    ["the rooster that crowed in the morn", "woke"],
    ["the farmer sowing his corn", "kept"],
    ["the horse and the hound and the horn", "belonged to"]
  ]

@doc """
  Return verses of the nursery rhyme 'This is the House that Jack Built'.
  """

  @spec recite(start :: integer, stop :: integer) :: String.t()
  def recite(1, 1) do
    "This is the house that Jack built.\n"
  end
  def recite(start, stop) when start === stop do
    "This is " <> build_line(start) <> "in the house that Jack built.\n"
  end
  def recite(start, stop) do
    Enum.map(start..stop, fn n -> recite(n, n) end)
    |> Enum.join()
  end

  defp build_line(n) when n > 1 do
    [actor, action] = Enum.at(@actors, n - 2)
    "#{actor} that #{action} #{build_line(n - 1)}"
  end
  defp build_line(1) do
    ""
  end
end
